README.md
# SauceLabs

saucelabs gem is written in simple ruby so as to be able to set the desired capabilities and browser options for selenium/watir grid execution
and also be able to dynamically set the browser choice through an environment variable and execute the same scripts across browsers.The gem was
initially written to help connect to saucelabs, however the current version support local execution, selenium grid execution and execution against
Sauce Labs cloud environment. Please bear in mind that to execute against Sauce Labs cloud, you would require to register and get a username, password
to access Sauce Labs on their [website] (http://www.saucelabs.com)
Also platform and device parameters in the desired capabilities object when creating a connection to
Sauce Labs. It can also be used to create a connection to local browser instance in addition to Sauce Labs
The gem has been tested on MRI ruby and JRuby. Please report any issues and I will try my best to fix asap.  

##Pre-Requisites
If you are planning to use a Selenium GRID or connect to Sauce Labs, it is assumed that you have an understanding on the Selenium GRID connection parameters. Sauce Labs works as a large Selenium GRID with username and password auth.

## Background

saucelabs gem is intended to be able to extract 4 main parameters viz. browser, version, platform and device
that gets set in the desired capabilities object. The gem creation got inspired by the below blog post
http://seleniumframework.wordpress.com/2014/05/18/pattern-for-running-multiple-cucumber-projects-on-ci-server/
Though the gem name is saucelabs , it can be used to create browser instances locally and against any selenium grid.

## Local Execution

````ruby
require 'saucelabs'
SauceLabs.watir_browser(browser = :firefox, browser_options = {})
# or
require 'saucelabs'
SauceLabs.selenium_driver(browser = :firefox, browser_options = {})
````
The methods use firefox as default browser and browser_options as optional parameter. You can specify other browsers
by the following:  

````ruby
SauceLabs.watir_browser(browser = :chrome, browser_options = {})
# or
SauceLabs.selenium_driver(browser = :chrome, browser_options = {})
````

## Executing against local selenium grid
It is assumed that you have started the grid as per the page [Grid 2] (http://code.google.com/p/selenium/wiki/Grid2)

Another example of being able to use chrome 35 on Windows8 platform  

````ruby
SauceLabs.watir_browser(browser = :'chrome35|windows8', browser_options = {:url => 'http://localhost:4444/wd/hub'})
# or
SauceLabs.selenium_driver(browser = :'chrome35|windows8', browser_options = {:url => 'http://localhost:4444/wd/hub'})
````

## Executing against Sauce Labs

Uses chrome 35 on Windows8 platform  

````ruby
SauceLabs.watir_browser(browser = :'chrome35|windows8', browser_options = {:url => 'http://username:password@ip:port/wd/hub'})
# or
SauceLabs.selenium_driver(browser = :'chrome35|windows8', browser_options = {:url => 'http://username:password@ip:port/wd/hub'})
````

## More ways and parameters (using conf)

Another example of being able to use chrome 35 on Windows8 platform (The :url can also be set using SauceLabs.conf, see below)  

````ruby
SauceLabs.watir_browser(browser = :'chrome35|windows8', browser_options = {})
# or
SauceLabs.selenium_driver(browser = :'chrome35|windows8', browser_options = {})
````


Yet Another example of being able to use chrome on mac platform and Android device (The :url needs to be set using SauceLabs.conf, see below)  

````ruby
SauceLabs.watir_browser(browser = :'chrome|linux|android', browser_options = {})
# or
SauceLabs.selenium_driver(browser = :'chrome|linux|android', browser_options = {})
````

Yet Another example of being able to use safari on mac platform and iphone device (The :url needs to be set using SauceLabs.conf, see below)  

````ruby
SauceLabs.watir_browser(browser = :'chrome|linux|iphone', browser_options = {})
# or
SauceLabs.selenium_driver(browser = :'chrome|linux|iphone', browser_options = {})
````

A simple ruby code example with watir and executing locally:

````ruby
require 'saucelabs'
include SauceLabs

ENV['BROWSER'] = "chrome"
@browser = SauceLabs.watir_browser
@browser.goto "http://www.google.com"
@browser.quit
````

A simple ruby code example with selenium and executing locally:

````ruby
require 'saucelabs'
include SauceLabs

ENV['BROWSER'] = "chrome"
@browser = SauceLabs.selenium_driver
@browser.navigate.to "http://www.google.com"
@browser.quit
````

A simple ruby code example with watir and executes against saucelabs:

````ruby
require 'saucelabs'
include SauceLabs

@browser = SauceLabs.watir_browser(:'chrome35|windows8', {:url => 'http://username:password@ip:port/wd/hub'})
@browser.goto "http://www.google.com"
@browser.quit
````

A simple ruby code example with watir and executes against local selenium grid:
It is assumed that you have started the grid as per the page [Grid 2] (http://code.google.com/p/selenium/wiki/Grid2)

````ruby
require 'saucelabs'
include SauceLabs

@browser = SauceLabs.watir_browser(:'chrome35|windows8', {:url => 'http://localhost:4444/wd/hub'})
@browser.goto "http://www.google.com"
@browser.quit
````


### Set BROWSER environment variable

In the above section, we have seen that it is very easy to set browser as a symbol
to the methods watir_browser and selenium_driver. However we would most likely want to dynamically set
the browser value. saucelabs reads from an environment variable 'BROWSER'. The following examples would help
understand the patterns. We can either use cucumber.yml file or set it on the command line as per your choice

````yml
default: BROWSER=firefox --format pretty --color
chrome: BROWSER=chrome --format pretty --color
ie: BROWSER=ie --format pretty --color
````

Execution from command line is simple too:

````
cucumber BROWSER=chrome features
````

The string pattern that is followed by this gem is as follows (Observer the '|' delimiter):


````
chrome36|win81
safari|linux|iphone
chrome|linux|android
````

Hence the following also applies:

````yml
default: BROWSER=chrome36|win81 --format pretty --color
safari_mac: BROWSER=safari|linux|iphone --format pretty --color
chrome_android: BROWSER=chrome|linux|android --format pretty --color
````



The first part determines the browser and its version
The second part determines the platform value that is set in capabilities object
The third part (used only for mobile device testing) determines the device



The following table is a mapping of all the platforms available on Sauce Labs website

| platform | description |
| --- | --- |
| windowsxp | Windows XP |
| osx109 | OS X 10.9 |
| osx106 | OS X 10.6 |
| osx108 | OS X 10.8 |
| linux | Linux |
| windows81 | Windows 8.1 |
| windows8 | Windows 8 |
| windows7 | Windows 7 |

The platform values can also be obtained from https://saucelabs.com/platforms

Additional platforms that are NOT on saucelabs however added to this gem to support selenium grid platforms

| platform | description |
| --- | --- |
| xp | XP|
| any | ANY |
| mac | MAC |
| unix | UNIX |
| vista | VISTA |
| windows | WINDOWS |



The following table is a mapping of all the devices available on Sauce Labs website

| device | description |
| --- | --- |
| iphone | iPhone |
| android | Android |


### Setting parameters using a block

We can also set parameters like sauce url, http_persistent and browser options. Below are some examples:


````ruby
SauceLabs.conf do | conf|
  conf.url = 'http://username:password@ip:port/wd/hub' #Saucelabs connection url
  conf.persistent_http = true
  conf.browser_options = { :switches => %w[--disable-extensions],:chromeOptions => {"args" => ["test-type" ]}}
end
````

## Documentation

You can view the RDocs for this project [here] (http://rubydoc.info/github/machzqcq/saucelabs/master/frames).

## Installation

Add this line to your application's Gemfile:

    gem 'saucelabs'

And then execute:

    $ bundle

Require the module:

    require 'saucelabs'

Or install it yourself as:

    $ gem install saucelabs


## Contributing

1. Fork it ( https://github.com/[my-github-username]/rd/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request


ChangeLog
=== Version 0.2


spec/spec_helper.rb
require 'saucelabs'
require 'selenium-webdriver'
require 'watir-webdriver'

spec/lib/saucelabs_spec.rb
require 'spec_helper'

describe "Example browser, platform and device combinations" do

  let(:watir_browser) { Watir::Browser }
  let(:selenium_driver) { Selenium::WebDriver}
  let(:capabilities) { Selenium::WebDriver::Remote::Capabilities }
  let(:desired_capabilities) { double('capabilities') }

  context "Firefox" do

    before(:each) do
      allow(capabilities).to receive(:firefox).and_return(desired_capabilities)
    end

    it "allows Firefox with no version specified" do
      expect(watir_browser).to receive(:new).with(:firefox)
      SauceLabs.watir_browser :firefox
    end

    it "allows Firefox 3" do
      expect(desired_capabilities).to receive(:version=).with('3')
      expect(watir_browser).to receive(:new).with(:remote, url: 'http://username:password@ip:port/wd/hub', desired_capabilities: anything())
      SauceLabs.watir_browser :firefox3, url: 'http://username:password@ip:port/wd/hub'
    end

    it "allows Firefox 6" do
      expect(desired_capabilities).to receive(:version=).with('6')
      expect(watir_browser).to receive(:new).with(:remote, url: 'http://username:password@ip:port/wd/hub', desired_capabilities: anything())
      SauceLabs.watir_browser :firefox6, url: 'http://username:password@ip:port/wd/hub'
    end

    it "allows Firefox 10" do
      expect(desired_capabilities).to receive(:version=).with('10')
      expect(watir_browser).to receive(:new).with(:remote, url: 'http://username:password@ip:port/wd/hub', desired_capabilities: anything())
      SauceLabs.watir_browser :firefox10, url: 'http://username:password@ip:port/wd/hub'
    end

    it "allows Firefox 25" do
      expect(desired_capabilities).to receive(:version=).with('25')
      expect(watir_browser).to receive(:new).with(:remote, url: 'http://username:password@ip:port/wd/hub', desired_capabilities: anything())
      SauceLabs.watir_browser :firefox25, url: 'http://username:password@ip:port/wd/hub'
    end

    it "allows Firefox 28" do
      expect(desired_capabilities).to receive(:version=).with('28')
      expect(watir_browser).to receive(:new).with(:remote, url: 'http://username:password@ip:port/wd/hub', desired_capabilities: anything())
      SauceLabs.watir_browser :firefox28, url: 'http://username:password@ip:port/wd/hub'
    end

    it "allows Firefox 29" do
      expect(desired_capabilities).to receive(:version=).with('29')
      expect(watir_browser).to receive(:new).with(:remote, url: 'http://username:password@ip:port/wd/hub', desired_capabilities: anything())
      SauceLabs.watir_browser :firefox29, url: 'http://username:password@ip:port/wd/hub'
    end

    it "allows Firefox 30" do
      expect(desired_capabilities).to receive(:version=).with('30')
      expect(watir_browser).to receive(:new).with(:remote, url: 'http://username:password@ip:port/wd/hub', desired_capabilities: anything())
      SauceLabs.watir_browser :firefox30, url: 'http://username:password@ip:port/wd/hub'
    end

  end

  context "Chrome" do

    before(:each) do
      allow(capabilities).to receive(:chrome).and_return(desired_capabilities)
    end

    it "allows Chrome with no version specified" do
      expect(watir_browser).to receive(:new).with(:chrome)
      SauceLabs.watir_browser :chrome
    end

    it "allows Chrome 26" do
      expect(desired_capabilities).to receive(:version=).with('26')
      expect(watir_browser).to receive(:new).with(:remote, url: 'http://username:password@ip:port/wd/hub', desired_capabilities: anything())
      SauceLabs.watir_browser :chrome26, url: 'http://username:password@ip:port/wd/hub'
    end

    it "allows Chrome 27" do
      expect(desired_capabilities).to receive(:version=).with('27')
      expect(watir_browser).to receive(:new).with(:remote, url: 'http://username:password@ip:port/wd/hub', desired_capabilities: anything())
      SauceLabs.watir_browser :chrome27, url: 'http://username:password@ip:port/wd/hub'
    end

    it "allows Chrome 28" do
      expect(desired_capabilities).to receive(:version=).with('28')
      expect(watir_browser).to receive(:new).with(:remote, url: 'http://username:password@ip:port/wd/hub', desired_capabilities: anything())
      SauceLabs.watir_browser :chrome28, url: 'http://username:password@ip:port/wd/hub'
    end

    it "allows Chrome 29" do
      expect(desired_capabilities).to receive(:version=).with('29')
      expect(watir_browser).to receive(:new).with(:remote, url: 'http://username:password@ip:port/wd/hub', desired_capabilities: anything())
      SauceLabs.watir_browser :chrome29, url: 'http://username:password@ip:port/wd/hub'
    end

    it "allows Chrome 30" do
      expect(desired_capabilities).to receive(:version=).with('30')
      expect(watir_browser).to receive(:new).with(:remote, url: 'http://username:password@ip:port/wd/hub', desired_capabilities: anything())
      SauceLabs.watir_browser :chrome30, url: 'http://username:password@ip:port/wd/hub'
    end

    it "allows Chrome 32" do
      expect(desired_capabilities).to receive(:version=).with('32')
      expect(watir_browser).to receive(:new).with(:remote, url: 'http://username:password@ip:port/wd/hub', desired_capabilities: anything())
      SauceLabs.watir_browser :chrome32, url: 'http://username:password@ip:port/wd/hub'
    end

    it "allows Chrome 36" do
      expect(desired_capabilities).to receive(:version=).with('36')
      expect(watir_browser).to receive(:new).with(:remote, url: 'http://username:password@ip:port/wd/hub', desired_capabilities: anything())
      SauceLabs.watir_browser :chrome36, url: 'http://username:password@ip:port/wd/hub'
    end

end
    context "Platforms Windows" do
      before(:each) do
        allow(capabilities).to receive(:chrome).and_return(desired_capabilities)
      end

      it "extracts Windows XP" do
        expect(desired_capabilities).to receive(:version=).with('26')
        expect(desired_capabilities).to receive(:platform=).with('Windows XP')
        expect(watir_browser).to receive(:new)
        SauceLabs.watir_browser :'chrome26|windowsxp', url: 'http://username:password@ip:port/wd/hub'
      end

      it "extracts Windows 8.1" do
        expect(desired_capabilities).to receive(:version=).with('35')
        expect(desired_capabilities).to receive(:platform=).with('Windows 8.1')
        expect(watir_browser).to receive(:new)
        SauceLabs.watir_browser :'chrome35|windows81', url: 'http://username:password@ip:port/wd/hub'
      end

      it "extracts Windows 7" do
        expect(desired_capabilities).to receive(:version=).with('33')
        expect(desired_capabilities).to receive(:platform=).with('Windows 7')
        expect(watir_browser).to receive(:new)
        SauceLabs.watir_browser :'chrome33|windows7', url: 'http://username:password@ip:port/wd/hub'
      end

      it "extracts Windows 8" do
        expect(desired_capabilities).to receive(:version=).with('36')
        expect(desired_capabilities).to receive(:platform=).with('Windows 8')
        expect(watir_browser).to receive(:new)
        SauceLabs.watir_browser :'chrome36|windows8', url: 'http://username:password@ip:port/wd/hub'
      end
    end


  context "Platforms OS X" do
    before(:each) do
      allow(capabilities).to receive(:safari).and_return(desired_capabilities)
    end

    it "extracts OS X 10.9" do
      expect(desired_capabilities).to receive(:version=).with('7')
      expect(desired_capabilities).to receive(:platform=).with('OS X 10.9')
      expect(watir_browser).to receive(:new)
      SauceLabs.watir_browser :'safari7|osx109', url: 'http://username:password@ip:port/wd/hub'
    end

    it "extracts OS X 10.6" do
      expect(desired_capabilities).to receive(:version=).with('6')
      expect(desired_capabilities).to receive(:platform=).with('OS X 10.6')
      expect(watir_browser).to receive(:new)
      SauceLabs.watir_browser :'safari6|osx106', url: 'http://username:password@ip:port/wd/hub'
    end

    it "extracts OS X 10.8" do
      expect(desired_capabilities).to receive(:version=).with('5')
      expect(desired_capabilities).to receive(:platform=).with('OS X 10.8')
      expect(watir_browser).to receive(:new)
      SauceLabs.watir_browser :'safari5|osx108', url: 'http://username:password@ip:port/wd/hub'
    end

end

  context "Platforms Linux" do
    before(:each) do
      allow(capabilities).to receive(:chrome).and_return(desired_capabilities)
    end

    it "extracts Linux" do
      expect(desired_capabilities).to receive(:version=).with('36')
      expect(desired_capabilities).to receive(:platform=).with('Linux')
      expect(watir_browser).to receive(:new)
      SauceLabs.watir_browser :'chrome36|linux', url: 'http://username:password@ip:port/wd/hub'
    end
  end

  context "Extract device" do
      it "extracts iPhone" do
      allow(capabilities).to receive(:chrome).and_return(desired_capabilities)
      expect(desired_capabilities).to receive(:platform=).with('Linux')
      expect(desired_capabilities).to receive(:device=).with('iPhone')
      expect(watir_browser).to receive(:new)
      SauceLabs.watir_browser :'chrome|linux|iphone', url: 'http://username:password@ip:port/wd/hub'
    end

    it "extracts Android" do
      allow(capabilities).to receive(:chrome).and_return(desired_capabilities)
      expect(desired_capabilities).to receive(:platform=).with('Linux')
      expect(desired_capabilities).to receive(:device=).with('Android')
      expect(selenium_driver).to receive(:for)
      SauceLabs.selenium_driver :'chrome|linux|android', url: 'http://username:password@ip:port/wd/hub'
    end
  end

end

spec/lib/parsed_values_spec.rb
require 'spec_helper'

class TestParsedValues
  include SauceLabs::ParsedValues
end


describe SauceLabs::ParsedValues do

  let(:parsed_values) { TestParsedValues.new }

  it "should extract browser name" do
    browser, version, platform,device = parsed_values.extract_values_from(:chrome)
    expect(browser).to eql :chrome
  end

  it "should extract browser name if browser and version are specified" do
    browser, version, platform,device = parsed_values.extract_values_from(:chrome35)
    expect(browser).to eql :chrome
  end

  it "should extract version if version is specified along with browser" do
    browser,version,platform,device = parsed_values.extract_values_from(:chrome35)
    expect(version).to eql '35'
  end

  it "should extract platform if platform is specified along with browser" do
    browser, version, platform,device = parsed_values.extract_values_from(:'chrome35|windows81')
    expect(platform).to eql 'Windows 8.1'
  end

  it "should extract iphone device if iphone is specified along with platform and browser" do
    browser, version, platform,device = parsed_values.extract_values_from(:'chrome35|windows81|iphone')
    expect(device).to eql 'iPhone'
  end

  it "should extract android device if android is specified along with platform and browser" do
    browser, version, platform,device = parsed_values.extract_values_from(:'chrome35|windows81|android')
    expect(device).to eql 'Android'
  end

  it "should report platform nil if platform is not defined in sauce_platforms" do
    browser, version, platform,device = parsed_values.extract_values_from(:'chrome35|windows82|android')
    expect(platform).to be_nil
  end

  it "should report device nil if device is not defined in sauce_devices" do
    browser, version, platform,device = parsed_values.extract_values_from(:'chrome35|windows82|xyz')
    expect(device).to be_nil
  end

end

lib/saucelabs.rb
require 'saucelabs/sauce_browser_factory'
require 'saucelabs/parsed_values'

module SauceLabs
  #attr_accessor :factory

  #
  # Creates a watir browser session and returns the browser object
  #
  # @example
  # SauceLabs.watir_browser(browser = :chrome, browser_options = {})
  # @param  [String] the browser string passed into the method
  # @param [Hash] the optional hash to specify browser options
  # @return [Object] browser session
  #

  def self.watir_browser(browser = :firefox, browser_options={})
    factory.watir_browser(browser,browser_options)
  end

  #
  # Creates a Selenium driver session and returns the driver object
  #
  # @example
  # SauceLabs.selenium_driver(browser = :chrome, browser_options = {})
  # @param  [String] the browser string passed into the method
  # @param [Hash] the optional hash to specify browser options
  # @return [Object] browser session
  #

  def self.selenium_driver(browser = :firefox, browser_options={})
    factory.selenium_driver(browser,browser_options)
  end

  #
  # Yields to Factory object if parameters are set using conf
  #
  # @return [Object] browser session
  #

  def self.conf
    yield factory
  end

  private

  def self.factory
    @factory ||= SauceBrowserFactory.new
  end

  

end

lib/saucelabs/sauce_browser_factory.rb
require_relative 'parsed_values'
require 'selenium-webdriver'
require 'watir-webdriver'

module SauceLabs
  #
  # This class has the code necessary to create an instance of browser after making a connection to saucelabs.
  # This class can also be used to create browser instance for local connections
  #
  class SauceBrowserFactory
    include ParsedValues

    attr_accessor :url,:options,:persistent_http

    #
    # Creates a watir browser session and returns the browser object
    #
    # @example
    # SauceLabs.watir_browser(browser = :chrome, browser_options = {})
    # @param  [String] the browser string passed into the method
    # @param [Hash] the optional hash to specify browser options
    # @return [Object] browser session
    #

    def watir_browser(browser,browser_options)
      target,options = browser_caps(browser,browser_options)
      create_watir_browser(target,options)
    end

    #
    # Creates a Selenium driver session and returns the driver object
    #
    # @example
    # SauceLabs.selenium_driver(browser = :chrome, browser_options = {})
    # @param  [String] the browser string passed into the method
    # @param [Hash] the optional hash to specify browser options
    # @return [Object] browser session
    #

    def selenium_driver(browser,browser_options)
      target,options = browser_caps(browser,browser_options)
      create_selenium_driver(target,options)
    end

    private

    def create_watir_browser(target,options)
      if options.empty?
        Watir::Browser.new target
      else
        Watir::Browser.new target,options
      end
    end

    def create_selenium_driver(target,options)
      if options.empty?
        Selenium::WebDriver.for target
      else
        Selenium::WebDriver.for target,options
      end
    end

    #
    # Returns the target and options including the capabilities
    #
    # @param  [String] the browser string passed into the method
    # @param [Hash] the optional hash to specify browser options
    # @return [Symbol,Hash] browser as symbol and options as Hash
    #

    def browser_caps(browser,browser_options)
      target = (browser.to_sym if ENV['BROWSER'].nil? or ENV['browser'].empty?) || (ENV['BROWSER'].to_sym)
      browser,version,platform,device = extract_values_from(target)
      options = {}
      options.merge! browser_options
      caps = capabilities(browser,version,platform,device)
      options[:url] = url if url
      if options.include? :url
        browser = :remote
        options[:desired_capabilities] = caps
      end
      options[:http_client] = http_client if persistent_http or options.delete(:persistent_http)
      return browser,options
    end


    def capabilities(browser, version, platform,device={})
      capabilities = Selenium::WebDriver::Remote::Capabilities.send browser
      capabilities.version = version unless version.nil? or version.empty?
      capabilities.platform = platform unless platform.nil? or platform.empty?
      capabilities.device = device unless device.nil? or device.empty?
      capabilities
    end

    #
    # Returns a persistent http connection object
    #
    # @return [Object] Persistent http connection object
    #

    def http_client
     client = Selenium::WebDriver::Remote::Http::Persistent.new
     #client.proxy_uri="http://abc"
     #client.read_timeout = 180
     #client.socket_options = {}
     client
    end


  end

end

lib/saucelabs/parsed_values.rb
module SauceLabs
  module ParsedValues

    #
    # Extracts browser, version, platform, device from the browser string
    #
    # @example
    #   extract_values_from(:'safari5|linux|iphone') will extract
    # browser = safari
    # version=5
    # platform=Linux
    # device=iPhone
    #
    # @param  [String] the browser string passed into the method
    # @return [String,String,String,String] browser, version, platform and device
    #
    def extract_values_from(browser_string)
      browser = extract_browser(browser_string).to_sym
      version = extract_version(browser_string)
      platform = extract_platform(browser_string)
      device = extract_device(browser_string)
      return browser,version,platform,device
    end

    private

    #
    # Extracts browser from browser string
    #
    # @example
    #   extract_browser(:'safari5|linux|iphone') will extract
    # browser = safari
    #
    # @param  [String] the browser string passed into the method
    # @return [String] the browser value
    #
    def extract_browser(value)
      browser = value.to_s.split(/\d+/)[0]
      browser = browser.to_s.split('|')[0] if browser.to_s.include? '|'
      browser
    end

    #
    # Extracts version from browser string
    #
    # @example
    #   extract_version(:'safari5|linux|iphone') will extract
    # browser = safari
    #
    # @param  [String] the browser string passed into the method
    # @return [String] the version value
    #
    def extract_version(value)
      value = value.to_s.split('|')[0] if value.to_s.include? '|'
      regexp_to_match = /\d{1,}/
      if (not regexp_to_match.match(value).nil?)
        version = regexp_to_match.match(value)[0]
      else
        version = ''
      end
      version
    end

    #
    # Extracts platform from browser string
    #
    # @example
    #   extract_platform(:'safari5|linux|iphone') will extract
    # browser = safari
    #
    # @param  [String] the browser string passed into the method
    # @return [String] the platform value
    #
    def extract_platform(value)
      platform = value.to_s.split('|')[1] if value.to_s.include? '|'
      sauce_platforms
      @sauce_platforms[platform] if not platform.nil?
    end

    #
    # Extracts device from browser string
    #
    # @example
    #   extract_device(:'safari5|linux|iphone') will extract
    # browser = safari
    #
    # @param  [String] the browser string passed into the method
    # @return [String] the device value
    #

    def extract_device(value)
      device = value.to_s.split('|')[2] if value.to_s.include? '|'
      sauce_devices
      @sauce_devices[device] if not device.nil?
    end

    #
    # Defines the platforms available on Sauce Labs as a Hash

    def sauce_platforms
      @sauce_platforms = Hash.new
      @sauce_platforms["windowsxp"] = "Windows XP"
      @sauce_platforms["osx109"] = "OS X 10.9"
      @sauce_platforms["osx106"] = "OS X 10.6"
      @sauce_platforms["osx108"] = "OS X 10.8"
      @sauce_platforms["linux"] = "Linux"
      @sauce_platforms["windows81"] = "Windows 8.1"
      @sauce_platforms["windows8"] = "Windows 8"
      @sauce_platforms["windows7"] = "Windows 7"
      @sauce_platforms["any"] = "ANY"
      @sauce_platforms["mac"] = "MAC"
      @sauce_platforms["unix"] = "UNIX"
      @sauce_platforms["vista"] = "VISTA"
      @sauce_platforms["windows"] = "WINDOWS"
      @sauce_platforms["xp"] = "XP"
    end

    #
    # Defines the devices available on Sauce Labs as a Hash


    def sauce_devices
      @sauce_devices = Hash.new
      @sauce_devices["iphone"] = "iPhone"
      @sauce_devices["android"] = "Android"
    end

  end
end